### PR TITLE
max including dust + unsupported contract estimation

### PR DESCRIPTION
**Pull request by rolandpo**
Changes for including all decimal places when selecting max balance on swap page for issue #652, and estimate display change for issue #653.
The max balance code might seem a bit overly complex, but I think it's needed to cover all the weird formatting edge cases.

### CODE REVIEW

**Comment by rolandpo** [comment]
also added a change to align the balance header in the history tab to be consistent with the swap page

**Review by sparrowDom — COMMENTED** [review]
I think the logic to figure out if max balance is currently set is quite complicated. The more complicated the logic the better chance we will introduce errors to it.  

Also If I input manually the maximum amount the UI doesn't figure out to treat it as a max value. 

All coins we are displaying are stablecoins. I would suggest that we should figure out solely from the value of the input if that input value should be treated as a maxValue one. I was so before, curious what were the edge cases that approach failed to cover?

File: dapp/src/components/buySell/SwapCurrencyPill.js [hunk 1]
```diff
@@ -418,16 +429,24 @@ const SwapCurrencyPill = ({
...
                     onAmountChange(valueNoCommas)
                   }
                 }}
                 onBlur={(e) => {
+                  setMaxBalanceSet(
+                    displayBalance &&
+                      Number(displayBalance.detailedBalance) ===
+                        Number(coinValue)
+                  )
                   const valueRounded = removeCommas(
```
Comment (sparrowDom) [inline dapp/src/components/buySell/SwapCurrencyPill.js:450]: Also `onBlur` still seems to set the value of the swap calling `onAmountChange` with the rounded amount. So by: 
- clicking on `Max` button
- focusing into the input field
- focusing out (by clicking `Swap` lets say)
- the `onBlur` will be called, by rounding the amount and leaving the `dust` in the user's wallet

File: dapp/src/components/buySell/SwapCurrencyPill.js [hunk 2]
```diff
@@ -418,16 +429,24 @@ const SwapCurrencyPill = ({
             {topItem && (
               <input
                 type="text"
-                value={coinValue}
+                value={formatMax(coinValue)}
                 placeholder="0.00"
                 onChange={(e) => {
+                  console.log('change')
```
[the remark this answers is not in the document]
Comment (rolandpo) [reply dapp/src/components/buySell/SwapCurrencyPill.js]: yep

File: dapp/src/components/buySell/SwapCurrencyPill.js (hunk 1 above)
Comment (rolandpo) [reply dapp/src/components/buySell/SwapCurrencyPill.js:454]: fixed

**Comment by rolandpo** [comment]
I think the idea of swapping one value but displaying a different one is what makes this a bit complicated. Currently when you input a value that happens to be max, then it will be recognised only on onBlur.

Otherwise, if my max is 10, then 10.00 will display as max, but I wouldn't be able to remove trailing zeroes on edit because the dust value gets locked between 2-6 decimals when passed to onChange.

**Comment by rolandpo** [comment]
So to ensure that input editing isn't affected in any way, I'm only asserting maxValueSet either when 'max' button is clicked or on onBlur. And then deasserting whenever the value get edited again (even if it doesn't change the actual value like a trailing zero)

**Comment by sparrowDom** [comment]
Aha I see thanks for explaining the rationale behind it. 

The one nitpick issue I still experience is if: 
- click on the max value, input changes to balance of coin. E.g. `1.981981`
- change value of coin to something else say `1`
- copy paste back the max value of coin: `1.981981`

The `Max` link is still present on the left, but it should be hidden

**Comment by sparrowDom** [comment]
Let me take a stab at this to see if some code simplification is possible

**Comment by rolandpo** [comment]
> * click on the max value, input changes to balance of coin. E.g. `1.981981`
> * change value of coin to something else say `1`
> * copy paste back the max value of coin: `1.981981`
> 
> The `Max` link is still present on the left, but it should be hidden

That would be consistent no? Either balance is exactly 1.981981 and on pasting the value and clicking off the field the Max link hides, or the exact balance has some extra dust and a copy paste only takes up to 6 decimal places so the Max link stays present

**Comment by sparrowDom** [comment]
@rolandpo have tried to simplify code somewhat. Can you check if this PR and see if it achieves all you wanted?
https://github.com/OriginProtocol/origin-dollar/pull/987

My though process was that lets allow the component to set the maximum amount of coins including dust and communicate that to the parent component (`SwapHomepage`). But only display the "nice" value that is floored to show maximum 6 decimals.

**Comment by sparrowDom** [comment]
> 

Figured later why copy/paste hasn't worked but works on current master. Dust usually has more than 6 decimals. And when copy pasting stuff the code needs to compare both amounts rounded to 6 decimals, since user isn't able to input more than 6

**Pull request merged**